File: src/csharp/30 - Substring with Concatenation of All Words.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace CSharpConsoleApp
{
    static class Helper
    {
        public static List<int> AllIndexOf(this string s, string value)
        {
            // 迭代器方式生成位置序列
            IEnumerable<int> InnerFunc0()
            {
                for (int i = s.IndexOf(value); i != -1; i = s.IndexOf(value, i+1))
                {
                    yield return i;
                }
            }

            // 尾递归方式将位置加入结果集中
            List<int> InnerFunc1(List<int> result, int start)
            {
                var index = s.IndexOf(value, start);
                if (index != -1)
                {
                    result.Add(index);
                    return InnerFunc(result, index + 1);
                }
                return result;
            }

            // return InnerFunc1(new List<int>(), 0);
            return new List<int>(InnerFunc0());
        }

        public static bool IsEmpty<TKey, TValue>(this Dictionary<TKey, TValue> dict) where TKey : notnull
            => dict.Count == 0;

        public static bool IsEmpty<T>(this ICollection<T> collection) => collection.Count == 0;

        public static string Format(this List<int> list)
        {
            if (list.IsEmpty())
            {
                return "[ ]";
            }
            return "[ " + list.Select(i => i.ToString()).Aggregate((s1, s2) => s1 + ", " + s2) + " ]";
        }
    }

    class Program
    {
        static bool IsExist(int index, Dictionary<string, int> count, Dictionary<string, HashSet<int>> dict)
        {
            if (count.IsEmpty())
            {
                return true;
            }
            foreach (var (word, indexes) in dict)
            {
                if (indexes.Contains(index) && count.ContainsKey(word))
                {
                    if (count[word] == 1)
                    {
                        count.Remove(word);
                    }
                    else
                    {
                        count[word] = count[word] - 1;
                    }
                    return IsExist(index + word.Length, count, dict);
                }
            }
            return false;
        }

        static List<int> FindSubstring(string s, string[] words)
        {
            var len = 0;
            var count = new Dictionary<string, int>();
            foreach (var word in words)
            {
                len += word.Length;
                count[word] = count.ContainsKey(word) ? (count[word] + 1) : 1;
            }

            var dict = new Dictionary<string, HashSet<int>>();
            foreach (var word in count.Keys)
            {
                dict[word] = new HashSet<int>(s.AllIndexOf(word));
                if (dict[word].IsEmpty())
                {
                    return new List<int>();
                }
            }

            var result = new List<int>();
            for (int i = 0; i < s.Length - len + 1; i++)
            {
                if (IsExist(i, new Dictionary<string, int>(count), dict))
                {
                    result.Add(i);
                }
            }
            return result;
        }

        static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");
            Console.WriteLine(FindSubstring("barfoothefoobarman", new string[] { "foo", "bar" }).Format());
            Console.WriteLine(FindSubstring("wordgoodgoodgoodbestword", new string[] { "word", "good", "best", "word" }).Format());
            Console.WriteLine(FindSubstring("barfoofoobarthefoobarman", new string[] { "bar", "foo", "the" }).Format());
            Console.WriteLine(FindSubstring("acaaa", new string[] { "aa", "ca" }).Format());
        }
    }
}

File: src/fsharp/30 - Substring with Concatenation of All Words.fs
open System

// 为String类型扩展方法
type String with
    // 尾递归函数实现
    member this.AllIndexOf0 (value: string) = 
        let rec func start list =
            match this.IndexOf(value, (start: int)) with
            | -1 -> list
            | i -> func (i+1) (i::list)
        func 0 []
    // 序列生成式实现
    member this.AllIndexOf1 (value: string) =
        seq {
            let mutable i = this.IndexOf(value)
            while i <> -1 do
                yield i
                i <- this.IndexOf(value, i+1)
        } |> List.ofSeq
    member this.AllIndexOf value = this.AllIndexOf1 value

let findSubString (s: string) (words: string array) =
    // 所有单词总长度
    let len = words |> Array.sumBy (fun word -> word.Length)
    // 各单词需要出现的次数
    let count = words |> Array.countBy id |> Map.ofArray
    // 单词及其在句子中出现的位置
    let map = words |> Array.map (fun word -> (word, s.AllIndexOf(word) |> Set.ofList)) |> Map.ofArray
    // 尾递归实现、测试指定位置是否能匹配一个目标词、直到“吃掉”所有词语、此时便能保证所有单词完成连接
    let rec exist index (count: Map<string, int>) =
        if count.IsEmpty then true else 
            let hitted = map |> Map.filter (fun word indexes -> indexes.Contains(index) && count.ContainsKey(word))
            if hitted.IsEmpty then false else 
                let word = hitted |> Map.toSeq |> Seq.head |> fst
                let next = if count.[word] = 1 then count.Remove(word) else count.Add (word, count.[word]-1)
                exist (index+word.Length) next
    seq { for i in 0 .. s.Length-len+1 do if exist i count then yield i } |> List.ofSeq

[<EntryPoint>]
let main argv =
    printfn "Hello World from F#!"
    //"barfoothefoobarman".AllIndexOf0 "foo" |> printfn "%A"
    //"barfoothefoobarman".AllIndexOf1 "foo" |> printfn "%A"
   
    ("barfoothefoobarman", [| "foo"; "bar" |]) ||> findSubString |> printfn "%A"
    ("wordgoodgoodgoodbestword", [| "word"; "good"; "best"; "word" |]) ||> findSubString |> printfn "%A"
    ("barfoofoobarthefoobarman", [| "bar"; "foo"; "the" |]) ||> findSubString |> printfn "%A"

    0 // return an integer exit code
